### PR TITLE
Add makeheaders

### PR DESCRIPTION
**Pull request by xoviat**

### CODE REVIEW

**Comment by xoviat** [comment]
Ping @conda-forge/staged-recipes

**Comment by dougalsutherland** [comment]
From the homepage that you're using:

> This is a mirror-repo of the famous makeheaders tool with various fixes and tweaks I made. The official upstream version is the "proper" version to use, and can be found at [their website](http://www.hwaci.com/sw/mkhdr/).
>
> [...]
>
> My main purpose for uploading this repo is to email the link to the upstream authors so they can integrate some or all of the changes if they want, which will hopefully make this mirror obsolete.

So this doesn't really seem like the right source to package, at least not under the name `makeheaders`.

**Comment by xoviat** [comment]
Let me rephrase: the original source is unmaintained. So if we use the original source, we are providing software with more bugs and a lower quality than this source. I personally don't see a reason to do that.

**Comment by dougalsutherland** [comment]
I don't know; maybe something like `makeheaders-rowanthorpe`, though that's kind of gross.

If this were a super-standard fork, that would be one thing. But the repo doesn't have any github stars, and @rowanthorpe explicitly says this isn't `the "proper" version to use`. So why should this one be named `makeheaders` and not [this one](https://github.com/TaniusTech/makeheaders), [this one](https://github.com/bubbapizza/makeheaders), or [this one](https://github.com/bjconlan/makeheaders)? Not saying that his version isn't currently the best fork or whatever, I don't know, but what if a better one comes along and uses a different/incompatible versioning scheme?

**Comment by xoviat** [comment]
> Not saying that his version isn't currently the best fork or whatever, I don't know, but what if a better one comes along and uses a different/incompatible versioning scheme?

I claim that it is the best fork. I also claim that a better fork will not come along because this fork is the only one with a merged pull request and a maintenance history.

**Comment by dougalsutherland** [comment]
Just for context on a brief search of what others have done:

- [Gentoo](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-util/makeheaders/makeheaders-0_p4.ebuild?id=6d3f0053feb20396daf71d11eac21459989a0ac2), the I-think-inherited-from-Gentoo [exherbo](http://git.exherbo.org/summer/packages/dev-util/makeheaders/index.html), and [AUR](https://aur.archlinux.org/packages/makeheaders/) packages use the original source
- Homebrew rejected a PR of a fork because "upstream doesn't really have a release process it seems and this is a fork of an unofficial github repo" (https://github.com/Homebrew/legacy-homebrew/pull/34319#issuecomment-63810246)

Didn't see any other packages of it.

@rowanthorpe, do you have opinions on this? If you wanted to call your fork `makeheaders-rt` or whatever and tag a release, then that would remove any doubt about packaging it. If you don't want to do that, maybe we can get an opinion from others on conda-forge about the best option here.

File: recipes/makeheaders/gh_1.patch [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@@ -1,2 +1,3 @@
+ /makeheaders
+ /makeheaders.c~
++build
+\ No newline at end of file
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+new file mode 100644
+index 0000000..492bad2
+--- /dev/null
++++ b/CMakeLists.txt
```
Comment (dougalsutherland) [inline recipes/makeheaders/gh_1.patch]: Since this patch was merged upstream (https://github.com/rowanthorpe/makeheaders/commit/5d50a25e12f771a8849eb7e189e4875387bf0dcf), probably drop it and just package a later commit.

File: recipes/makeheaders/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+source:
+  fn: makeheaders-{{ version }}.src.tar.xz
+  url: https://github.com/rowanthorpe/makeheaders/archive/{{ commit }}.tar.gz
+  sha256: {{ sha256 }}
+  patches:
+    - gh_1.patch
+
+build:
+  number: 0
+  skip: true  # [win and py<35]
```
Comment (dougalsutherland) [inline recipes/makeheaders/meta.yaml]: Should be `py<=35`? No need to build on both `py35` and `py36`. Also, does it not work on VC9 or do you just not want to bother?

File: recipes/makeheaders/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  features:
+    - vc14    # [win and py>=35]
+
+requirements:
+  build:
+    - toolchain
+    - cmake >=3.4.3
```
Comment (dougalsutherland) [inline recipes/makeheaders/meta.yaml]: The `CMakeLists` only says `3.0`; not a big deal, but is `3.4.3` actually needed?

File: recipes/makeheaders/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  requires:
+    - python {{ environ['PY_VER'] + '*' }}  # [win]
+    - cmake >=3.4.3
+
+  commands:
+    - makeheaders
+
+about:
+  home: https://github.com/rowanthorpe/makeheaders
+  license: MIT
```
Comment (dougalsutherland) [inline recipes/makeheaders/meta.yaml]: It's actually BSD-2-Clause. And I think we need a `license_file`; since the repo doesn't have one, you should maybe manually package the comment from [the source file](https://github.com/rowanthorpe/makeheaders/blob/master/makeheaders.c#L6).

**Comment by xoviat** [comment]
> @rowanthorpe, do you have opinions on this? If you wanted to call your fork makeheaders-rt or whatever and tag a release, then that would remove any doubt about packaging it. If you don't want to do that, maybe we can get an opinion from others on conda-forge about the best option here.

So I'm not going to spend time on changes until we can get agreement here. The reason that I added this is that it was useful tooling for working on NumPy. The alternative is that it isn't packaged and people have to spend time downloading and compiling it themselves. I am also skeptical that you will get a response from the author of that package.

**Comment by rowanthorpe** [comment]
> > @rowanthorpe, do you have opinions on this? If you wanted to call your fork makeheaders-rt or whatever and tag a release, then that would remove any doubt about packaging it. If you don't want to do that, maybe we can get an opinion from others on conda-forge about the best option here.
> 
> So I'm not going to spend time on changes until we can get agreement here. The reason that I added this is that it was useful tooling for working on NumPy. The alternative is that it isn't packaged and people have to spend time downloading and compiling it themselves. I am also skeptical that you will get a response from the author of that package.

While researching how to answer your questions I just discovered the following:

* The [documentation of the makeheaders embedded in Fossil](https://www.fossil-scm.org/index.html/artifact/c09a2e625c04fe6a) is subtly different from the [documentation of the original "upstream version"](http://www.hwaci.com/sw/mkhdr/makeheaders.html) that I was basing my repo on.
* The "history" section of that documentation for the version in Fossil mentions the following:
    > The makeheaders program was first written by D. Richard Hipp (also the original author of SQLite and Fossil) in 1993. Hipp open-sourced the project immediately, but it never caught on with any other developers and it continued to be used mostly by Hipp himself for over a decade. When Hipp was first writing the Fossil version control system in 2006 and 2007, he used makeheaders on that project to help simplify the source code. As the popularity of Fossil increased, the makeheaders that was incorporated into the Fossil source tree became the "official" makeheaders implementation.
* Then when I found the commit-history of Fossil's makeheaders I saw [changes up until March 2017](https://www.fossil-scm.org/index.html/finfo?name=src/makeheaders.c&m=ef97646ffc78b840)

[aaargh!]

So it seems:

* The "official" version of makeheaders according to [its author](http://www.hwaci.com/contact.html) (DRH) is now the one embedded in Fossil.
* He didn't update the separate "original upstream" makeheaders project to indicate as such, so I've only found that out now after deep-diving on the issue.

I think it is good to have an *independent*, *maintained* makeheaders in some form, and it seems DRH also wanted that originally but in the end just abandoned it in favour of the version embedded in Fossil, which is a project he actually got some uptake on. The unfortunate part is just that he didn't indicate that on the "original upstream" project (or its docs), and didn't remove it either.

Now I know that version exists I will merge the latest changes from there into my repo (possibly already fixing some of what my commits were for), and will try to contact him via the Fossil project's issue tracker (as the email I sent in June last year never got a response) to see if he is interested in re-externalizing makeheaders, and then just "depending" on it (or even copy-paste "depending") in Fossil. If so I will ask if he is happy to maintain it separately himself, or failing that if he'd like me to (makeheaders has a mostly static and well-defined use-case so shouldn't be a big maintenance-burden). If both of those options are no-go, then I will submit any of my non-redundant fixes/tweaks as patches to the Fossil version, and then perhaps I could keep the github repo as a lockstep clone of the code as it appears in the Fossil version. If for some reason he is ever averse to changes which most "external users" want but he doesn't want in the Fossil version, only then do I think it might be worth maintaining a "rename fork". BTW: if he is willing to maintain makeheaders separately himself again I am almost certain it won't be on Github, but instead using Fossil itself (even if just on principle, like the Common Lisp guru who gets fed up with always being asked "why don't you just use Clojure though?").

I'll keep you updated. In the meantime if you think I should approach it differently I am open to suggestions, please let me know. I will probably get to merging his changes in a day or two before contacting him, so if you let me know before then it will be in time.

**Comment by rowanthorpe** [comment]
Also: My repo was originally intended as a convenience for upstream to see the patches, so if the end-solution instead involves retaining my repo in some form, I will add the documentation html-page and a license-file to it.

**Comment by dougalsutherland** [comment]
Thanks @rowanthorpe! Though since there hasn't been any activity on the Fossil issue tracker in two years, maybe the mailing list would be a better way to get in touch.

A possible external version being maintained on Fossil is no problem as long as there's some place that conda-forge can download a tarball from. Or we can also grab only makeheaders out of the Fossil source and build that.

I think it's worth waiting a bit to see what the situation will be before making a decision about packaging here.
